go/libraries/doltcore/sqle/dprocedures/dolt_cherry_pick.go

```diff
@@ -0,0 +1,208 @@
+// or a commit without parent commit, performs merge and returns the new working set root value and
+// the commit message of cherry-picked commit as the commit message of the new commit created during this command.
+func cherryPick(ctx *sql.Context, dSess *dsess.DoltSession, roots doltdb.Roots, dbName, cherryStr string) (*doltdb.RootValue, string, error) {
+	// check for clean working set
+	headRootHash, err := roots.Head.HashOf()
+	if err != nil {
+		return nil, "", err
+	}
+
+	workingRootHash, err := roots.Working.HashOf()
+	if err != nil {
+		return nil, "", err
+	}
+	if workingRootHash != headRootHash {
+		return nil, "", ErrCherryPickUncommittedChanges
+	}
+
+	stagedRootHash, err := roots.Staged.HashOf()
+	if err != nil {
+		return nil, "", err
+	}
+	if stagedRootHash != headRootHash {
+		return nil, "", ErrCherryPickUncommittedChanges
+	}
+
+	doltDB, ok := dSess.GetDoltDB(ctx, dbName)
+	if !ok {
+		return nil, "", fmt.Errorf("failed to get DoltDB")
+	}
+
+	dbData, ok := dSess.GetDbData(ctx, dbName)
+	if !ok {
+		return nil, "", fmt.Errorf("failed to get dbData")
+	}
+
+	cherryCommitSpec, err := doltdb.NewCommitSpec(cherryStr)
+	if err != nil {
+		return nil, "", err
+	}
+	cherryCommit, err := doltDB.Resolve(context.TODO(), cherryCommitSpec, dbData.Rsr.CWBHeadRef())
+	if err != nil {
+		return nil, "", err
+	}
+
+	if len(cherryCommit.DatasParents()) > 1 {
+		return nil, "", fmt.Errorf("cherry-picking a merge commit is not supported")
+	}
+	if len(cherryCommit.DatasParents()) == 0 {
+		return nil, "", fmt.Errorf("cherry-picking a commit without parents is not supported")
+	}
+
+	cherryRoot, err := cherryCommit.GetRootValue(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	parentCommit, err := doltDB.ResolveParent(ctx, cherryCommit, 0)
+	if err != nil {
+		return nil, "", err
+	}
+	parentRoot, err := parentCommit.GetRootValue(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	dbState, ok, err := dSess.LookupDbState(ctx, dbName)
+	if err != nil {
+		return nil, "", err
+	} else if !ok {
+		return nil, "", sql.ErrDatabaseNotFound.New(dbName)
+	}
+
+	// use parent of cherry-pick as ancestor root to merge
+	mo := merge.MergeOpts{IsCherryPick: true}
+	mergedRoot, mergeStats, err := merge.MergeRoots(ctx, roots.Working, cherryRoot, parentRoot, cherryCommit, parentCommit, dbState.EditOpts(), mo)
+	if err != nil {
+		return nil, "", err
+	}
+
+	var tablesWithConflict []string
+	for tbl, stats := range mergeStats {
+		if stats.Conflicts > 0 {
+			tablesWithConflict = append(tablesWithConflict, tbl)
+		}
+	}
+
+	if len(tablesWithConflict) > 0 {
+		tblNames := strings.Join(tablesWithConflict, "', '")
+		return nil, "", fmt.Errorf("conflicts in table {'%s'}", tblNames)
+	}
+
+	workingRootHash, err = mergedRoot.HashOf()
+	if err != nil {
+		return nil, "", err
+	}
+
+	if headRootHash.Equal(workingRootHash) {
+		// TODO: should it be warning? or nothing?
```

Comment:
What does Git do in this case? Specifically... does it return a 0 exit code to indicate success? Does this only happen if the cherry-picked commit is an empty commit? 

The user requested to cherry-pick, so if it was applied successfully (even if nothing changed), then it seems like it's okay to return a success/0 exit code. Let me know if I'm just missing something about when this block would trigger.

Reply:
We return exit code 0 as nothing failed but just a message is printed saying nothing to commit in the CLI command. I was not sure whether to return 0 for success and return empty hash or the current head hash?

Reply:
Actually... checking out Git's behavior... if I try to cherrypick an empty commit or the current HEAD commit, I get a `1` exit code indicating the commit tree was not changed and a message about `nothing to commit, working tree clean`. What do you think about matching that behavior here?

Reply:
sounds good!